### DISCUSSION TITLE
[RFC] Official support for Svelte CSF

### DISCUSSION THREAD

**Discussion by JReinhold** [storybookjs/storybook#27092] [RFC]
### Summary

A majority of the Svelte community considers the Svelte CSF as implemented in [`@storybook/addon-svelte-csf`](https://github.com/storybookjs/addon-svelte-csf) a better way to write stories than with regular CSF. It’s closer to the syntax for writing regular Svelte components, it supports some features of Svelte that regular CSF doesn’t (such as [slots](https://svelte.dev/docs/special-elements#slot) and [snippets](https://svelte-5-preview.vercel.app/docs/snippets)), but it also comes with its own set of shortcomings.

This RFC proposes making Svelte CSF a first-class citizen of Storybook, including making it the official and default way to write stories for Svelte and documenting it in the docs.

As an open-source project, Storybook relies heavily on the community to build great things, and this is no exception. Given Storybook’s current roadmap, the core team can only participate in a guiding capacity, helping the community complete this effort with PR reviews and architectural guidance. This comes with its own set of challenges, but that’s the trade-off we have to make right now.

### Problem Statement

[Component Story Format (CSF)](https://storybook.js.org/docs/api/csf) is a portable standard for capturing UI component examples. It is based on [ECMAScript Modules (ESM)](https://tc39.es/ecma262/#sec-modules), where the default export contains metadata about the component being documented, and each named export captures a key state of that component.

This is great for JSX-based frameworks because JSX can be embedded in a regular ESM file and thus you can do everything with JSX components in CSF. However the same isn’t true for template-based frameworks like Svelte and Vue - you can’t define a Svelte component directly in an ESM-file, it has to be its own `.svelte` file. This limits what you can do in CSF without defining multiples of Svelte components in separate files just for the sake of writing stories.

Svelte provides [a set of APIs for mounting a component on the client](https://svelte.dev/docs/client-side-component-api) - this is the API Storybook uses internally to render CSF-based stories. However this API is limited, you can’t do the same things with root Svelte components as you can with components within other components. In practice, this means that you can’t define slots in regular CSF, because there’s no way to pass a [slot](https://svelte.dev/docs/special-elements#slot) to a component when using Svelte’s mounting API, and thus there’s no way for Storybook to take a user’s slot and pass it to the component. The upcoming Svelte 5 introduces [snippets](https://svelte-5-preview.vercel.app/docs/snippets) as a replacement for slots, which have the same limitations.

[`@storybook/addon-svelte-csf`](https://github.com/storybookjs/addon-svelte-csf) solves a majority of these problems by allowing users to write stories with the Svelte language (we call this “Svelte CSF”), that are then compiled behind the scenes to CSF that Storybook can understand. This is great, and even with the current lack of visibility, it’s still used in the majority of Svelte-based Storybooks today.

However the addon comes with its own set of shortcomings, such as

- Lack of general documentation
- No explicit documentation mapping CSF features to the Svelte CSF equivalent, nor any known workarounds listed
- Some features are not supported fully, such as the test runner or Portable Stories

Most prominently, there’s no clear guide on when to use Svelte CSF vs CSF, leaving newcomers confused and in the dark.

We want to solve this by making Svelte CSF a default, first-class citizen of Storybook.

### Non-goals

- This won’t solve templating languages in general in Storybook. This is specifically about Svelte.
- This won’t make Svelte CSF part of the CSF standard. We should attempt to make it possible for other CSF-based tools to support Svelte CSF, but we won’t make this part of regular CSF.
- This won’t support every single CSF feature 1-to-1 in Svelte CSF - at least not from day one. Some features don’t make sense in Svelte CSF, and some are harder to support than what can be achieved initially. We *should* however clearly identify which features are and aren’t supported.
- This won’t replace regular CSF for Svelte on a technical level. Storybook internals and tests rely heavily on regular CSF, and thus even the Svelte Storybook framework needs to support regular CSF still, although not necessarily in a publicly documented way.
- This won’t move the functionality of the addon into Storybook core, it will just ensure the addon is installed by default. We think it’s beneficial to keep the addon in its own repository with its own versioning. The Storybook monorepo can slow down development and the addon could benefit from higher velocity. This is not a final decision, so if anyone comes up with good counterarguments for why this should be baked into `@storybook/svelte` without the addon, we are open to discussing this.

### Implementation

To solve these problems we propose the following changes:

**1. Demo Project**

The purpose of this project is to identify and describe the differences between regular CSF and Svelte CSF. Ideally, it should have a set of stories that uses all the features of Storybook with regular CSF, and a companion Svelte CSF version of all stories demonstrating how to do the same with Svelte CSF, or how and why this is not possible at the moment.

This project should be a separate repository in the Storybook GitHub org.

See the Deliverables section for a list of features to demonstrate.

It’s important to re-iterate that the purpose of this is not to show off how great Svelte CSF is, it’s to teach users and maintainers the differences and learn where it falls short. Hopefully, any shortcomings should be addressable in follow-up work improving Svelte CSF.

**2. Document addon usage**

We need to clearly document how this addon is a special case of Storybook, to not cause any confusion. I’m not sure what shape this takes, it might not be more than a few sentences or paragraph, or something else completely.

**3. Convert all relevant code snippets in the docs to Svelte CSF**

We have 200+ code snippets in the docs that shows CSF usage, that we should attempt to convert to Svelte CSF. This is a big lift and will happen over a longer period. This is also where the community can help out the most.

As an example, here’s how [the first snippet in the Args doc](https://storybook.js.org/docs/writing-stories/args#story-args) could be converted:

```svelte
<!-- Button.stories.svelte -->
<script context="module">
  import Button from './Button.svelte';

  export const meta = {
    component: Button
  }
</script>

<script>
  import { Story } from '@storybook/addon-svelte-csf';
</script>

<Story name="Primary" args={{ primary: true, label: 'Button' }}/>
```

We should incrementally add Svelte CSF code snippets as the main snippets, and remove snippets for regular CSF when they are not needed anymore. There might be specific features where we’ll intentionally keep regular CSF docs around because Svelte CSF doesn’t support them.

**4. Install addon by default**

This might already be covered by https://github.com/storybookjs/storybook/pull/27070

**5. Automigration**

If the documentation only documents Svelte CSF, we should write an automigration that

1. Explains this shift from CSF to Svelte CSF, possibly pointing to a relevant docs section or blog post
2. Offers to add the addon and modify the `stories` glob to match

It should only run on projects that don’t already have the addon installed.

**6. Expose builder plugin**

The Svelte CSF Addon relies on a builder plugin (Webpack and Vite) to compile Svelte CSF files to regular CSF files virtually. Storybook is starting to support non-Storybook environments like [Portable Stories](https://storybook.js.org/docs/8.1/api/portable-stories-playwright), and for that to work with Svelte CSF we’d need to expose this builder plugin from the package so it can be added to any configuration necessary.

**7. Add stories and tests**

To confidently provide support for Svelte CSF we’d want to have tests and stories in the monorepo that ensures we don’t introduce any regressions. While the addon does have some unit testing and stories, we should be able to expand on those (perhaps based on the demo project described above), but we also want to add stories and tests directly to the Svelte renderer in the monorepo.

### Deliverables

> [!IMPORTANT]
> If you want to contribute to any of this, just ask questions, start doing the work or open pull requests to the relevant repositories (see [How to contribute](https://storybook.js.org/docs/8.1/contribute)). There’s no need to “reserve” a task or ask for being assigned to anything specifically before starting work on it. That system doesn’t work, often contributors reserving tasks ends up not doing them.

- [x]  Demo project - https://github.com/storybookjs/svelte-csf-demo
    - [x]  Scaffold repository with a `README` explaining the intent of the demo, how to use it and contribute @JReinhold 
    - [x]  Create an example PR for a feature that the community can use as a template for each feature demo. @JReinhold 
    - [x] See [the repos Overview for a list of deliverables](https://github.com/storybookjs/svelte-csf-demo#overview).
- [x]  Install addon by default with `storybook init`  https://github.com/storybookjs/storybook/pull/27070
- [ ]  Introductory documentation of Svelte CSF usage
- [ ] Convert docs code snippets to Svelte CSF (too many to list here)

### Risks

1. Changing the whole underlying syntax is a disruptive change that can confuse existing users that don’t know about Svelte CSF.
2. Only a subset of the deliverables gets done for a long time, leaving the overall Svelte experience in a worse state with half-documented features or functionality that is not well tested.
3. The Storybook core team doesn’t have the bandwidth to guide and review a high influx of Svelte community work, leaving everyone frustrated and work never merged.

### Unresolved Questions

- [ ] **Do we still need to support Webpack in Svelte projects?**

Over the past few years, the Svelte community has strongly favoured Vite over Webpack (See [npmtrends](https://npmtrends.com/@sveltejs/vite-plugin-svelte-vs-svelte-loader)). Adding official support for Svelte CSF is a major surface area increase, and it would be beneficial to keep that surface area under control by removing support for Svelte+Webpack. However, the Svelte core team still supports Webpack and we can’t ignore that we still have Svelte+Webpack users in Storybook - albeit a very small minority.

- [ ] **Should we release a major version of the addon before starting on docs?**

At the moment the community is trying to add support for Svelte 5 ([here](https://github.com/storybookjs/addon-svelte-csf/pull/177), and [here](https://github.com/tsar-boomba/addon-svelte-csf/pull/1)) in the addon. Nothing is set in stone yet but it looks like the best way forward with that would be a breaking change to the addon, only supporting Svelte 5, and using snippets instead of slots for defining stories. Furthermore the new syntax for defining meta ([old](https://github.com/storybookjs/addon-svelte-csf/blob/main/stories/button.stories.svelte) vs [new](https://github.com/storybookjs/addon-svelte-csf/blob/main/stories/metaexport.stories.svelte)) is better overall so we could consider removing support for the old syntax in that same major release.

If we did this, it would be best to do this before starting any docs work, given the syntax is slightly different. On the other hand, it would alienate all existing users that can’t/won’t upgrade to Svelte 5 for whatever reason. There isn’t a lot breaking in Svelte 5 and it still supports the Svelte 4 syntax, so requiring users to upgrade might not be the worst.

**Comment by xeho91** [comment]
> On the other hand, it would alienate all existing users that can’t/won’t upgrade to Svelte 5 for whatever reason. There isn’t a lot breaking in Svelte 5 and it still supports the Svelte 4 syntax, so requiring users to upgrade might not be the worst.

I'd like to put at ease regarding this matter, after what I found during my experiments.

While using an upgraded addon to support `svelte@5`, the custom _(user's)_ components which have used `svelte@4` features will still work. The addon's components _(`<Template />` & `<Story />`)_  got upgraded to `v5`, so basically what needs to be done is migrating from this:

```svelte
<Story let:args>
  <Button {...args} />
</Story>
```

to this:

```svelte
<Story>
  {#snippet children(args)}
    <Button {...args} />
  {/snippet}
</Story>
```

A little bit more verbose, yeah, but the benefits are that `args` can have full TypeScript support, and with snippets there's more flexibility.

~~In theory it might be possible to keep the previous syntax for backwards compatibility, as Svelte considered to allow using snippets and slots together. I haven't explored this to be honest.~~

**EDIT.** Nope, I was wrong. I get the following error:

```
Cannot use `<slot>` syntax and `{@render ...}` tags in the same component. Migrate towards `{@render ...}` tags completely. [slot_snippet_conflict]
```

**Reply by JReinhold** [reply]
I see about the component syntax, but I assume that the workspace's Svelte would still need to be upgraded to v5 right?

**Reply by xeho91** [reply]
> I see about the component syntax, but I assume that the workspace's Svelte would still need to be upgraded to v5 right?

If you mean `@storybook/svelte`, I am not entirely sure, I don't think so. 🤔 
I was able to make it work along with `@storybook/addon-svelte-csf` using svelte `v5` and updated addon's components.

I haven't fully grasped yet the backward compatibility. As far I can tell from the quick glimpse. If we'd rewrite `@storybook/svelte` [components](https://github.com/storybookjs/storybook/tree/next/code/renderers/svelte/src/components) to use svelte `v5` with runes and snippets, then it shouldn't break the components passed down by user which haven't been upgraded  yet.

**Reply by JReinhold** [reply]
What I mean is, the `svelte` dependency in the project's `package.json` would need to be `^5.0.0`, and not `^4.0.0` - since `svelte` is just a peer dependency of the Storybook Svelte integration.

While Svelte 5 itself supports Svelte 4 syntax in general, there are some breaking changes with Svelte 5 that could block project's from upgrading. Hopefully it shouldn't be often though.

**Reply by xeho91** [reply]
I think maintainers of the Svelte can give a better answer than I do. 🙏 
As far I can tell, we might not need to. But without any demo repository for reproduction and tests to see if anything breaks on `@storybook/addon-svelte-csf` _(after upgrading to upcoming next major version)_ or `@storybook/svelte` side, I can only speculate.

**Reply by benmccann** [reply]
Yes, if you want to use snippets or any new features from Svelte 5 then the user's project will need Svelte 5 installed and Storybook will need to list it as a peerDependency. I think it should be fine to increase the peerDependency version in the next major release of Storybook. It is intended to be fairly straightforward for users to upgrade their projects to Svelte 5. There are some breaking changes, but they're intended to be fairly minimal such that most users won't hit most of them: https://svelte-5-preview.vercel.app/docs/breaking-changes

**Reply by JReinhold** [reply]
Great, that matches my mental model. 

> It is intended to be fairly straightforward for users to upgrade their projects to Svelte 5. There are some breaking changes, but they're intended to be fairly minimal such that most users won't hit most of them

I think this is the key point.

**Comment by jacob-8** [comment]
@shilman, we discussed the idea of Kitbook shifting it's terminology from variants to stories, from `*.variants.ts` to `*.stories.ts` so that a simple Story in Kitbook would look like this:

```ts
import type { Story } from 'kitbook'
import type Component from './Greeting.svelte'

export const Simple: Story<Component> = {
  props: { // props is what we call these in Svelte, but Kitbook would also accepts `args` for the same purpose
    greeting: 'Hello World',
  },
  description: 'A simple greeting',
}
```

The comparable version in Storybook CSF (sans description) would look like this:

```ts
import type { Meta, StoryObj } from '@storybook/svelte'
import Greeting from './Greeting.svelte'
 
const meta: Meta<typeof Greeting> = {
  component: Greeting,
}
 
export default meta;
type Story = StoryObj<typeof meta>
 
export const Simple: Story = {
  args: {
    greeting: 'Hello World',
  },
}
```

Both files would be `Greeting.stories.ts` and thus if you were working in Storybook and wanted to run Kitbook, the basic function of mounting stories with props would just work. Obviously you'd lose any additional Storybook features that the user happened to implement. Conversely if you had a Kitbook and wanted to upgrade to Storybook (I call it an upgrade because the feature set and ecosystem is large) you could do so with minimal changes. At that point you could use either!

The purpose of all this is to help the wider dev community coalesce around some standards for component development. We want people to easily understand the usefulness of our tools and how they can improve their DX.

So my question then is, **how much of Storybook's Svelte community actually uses `*.stories.ts` files?** If it is ~40%, our alignment is still a win. But if you are pushing strongly for the Svelte CSF format to be the only encouraged method, our alignment has less purpose. I think there's major pros and cons to both using the traditional CSF format and the Svelte CSF so I'm just wanting to hear your thoughts on the balance between the two.

**Reply by shilman** [reply]
@jacob-8 Yeah I can see how things are a bit confusing!

#### What's the current status?

- Over the past few years, we've received tons of feedback from the Svelte community that they want to Svelte stories, both for ergonomics/familiarity and also for slot support, which has been missing in CSF.
- We've [figured out how to provide slot support in CSF](https://github.com/storybookjs/storybook/pull/28393), which addresses the biggest objection to CSF, at least to some degree. However, it does not address the ergonomics/familiarity issues.
- We are officially supporting Svelte Native CSF, and it's installed out of the box in new projects. However the sample stories are still CSF, so the majority of stories are likely still CSF. We don't collect telemetry about this.

#### Where are we headed? 

- At some point we'll switch the sample stories to `.stories.svelte` and update the documentation. I don't have a timeline, but hopefully soon.
- We'll continue to support "standard" CSF alongside Svelte Native CSF. We are working on a syntax update that should provide much better typesafety and autocompletion. I think this will be a pretty compelling iteration, along with Slot/Snippet support, but my guess is that the Svelte community will prefer the Svelte syntax.

Given how dynamic this is, we're not trying to make CSF an interoperable standard. Someday maybe, but not in the next 2y.

#### What does this mean for you?

We're trying to establish the terminology of "Story = Component + Inputs", which is used by Storybook and similar tools. So if you renamed Kitbook variants to stories, that would align with what we're doing.

As for being able to flexibly migrate between the two tools, it sounds great on the surface. Each tool will have its strengths and weaknesses, so being able to flexibly switch back and forth makes it possible to pick and choose (e.g. Kitbook for developing inside your app, storybook for CI).

My suggestion would be to wait a month or two for our CSF4 RFC and potentially join in the discussion there. It won't be the last major iteration, but it's a chance to align more closely AND also means that you won't make changes now and then have the rug pulled out from under you in a few months.

**Reply by jacob-8** [reply]
Wonderful! That is a very helpful clarification. Yes, I'm game to move forward with aligning terminology but not keeping hopes too high on interopability. And yes, please tag me in the CSF4 conversation as I will be unaware of it otherwise. My main contribution will be this:

In React, you can have 10 components in a file, therefore in `*.stories.ts` files you need to import and define which component you are mounting. In Svelte you only have 1 component per file and since the `*.stories.ts` file is colocated this is unnecessary boilerplate. Therefore we should be able to do this:

```ts
import type { Story } from '@storybook/svelte'
import type Component from './Greeting.svelte'

export const Simple: Story<Component> = {
  args: {
    greeting: 'Hello World',
  },
}
```

---

As an aside,  I'm sure you aware that Rich is pushing Svelte users towards the practice of everything coming through props (what you call args). Values, events, and snippets. I already commented about snippets (below), but this also makes things like logging events quite easy. If you provide a simple helper ([demo](https://svelte-5-preview.vercel.app/#H4sIAAAAAAAACo2RwWqDQBCGX2VYCiox8dSLUaGUQo-59RBDsDqm0nV2WceUIvvuZY1NGkKkl4WZ-b_Zf2YGUTcSOxFvB0FFiyIWT1qLUPC3dkF3RMkoQtGp3pQuk3SlaTRnOeXctFoZhmfVakVI_Nbwx8sRiV8LqiSajVG6g9qoFrxVNC9bnb7y1jm51nVPJTeKQKrDHp3YH9-9cxnA4DQABrk3BH5hDl0AaQalok5JXEl18AfYXpBdDE4ENnCkzSmJLoNQMu9tnJVqWTCmw8WQN-W8wDpFhbeKKTcqoiwnEYpWVU3dYCViNj3a8Lz4eQ__vYlEhgEmZyFMBsBCCg_aNfKD9a9Oqy80rtBxweg_usr1Xt57ZkWgqJRN-ZkO_rjlqbs_8oHNpjiJTvLsLji5OYNTfAWa-_iILZc2W_4hNi4ZwzAW7Ty7WNhscWZvj7GzP9AG8cwRAwAA)) like this:

```ts
function log_event(event_name) {
  return (args) => console.log({ [event_name]: args })
}
```

The users can use it like:

```ts
import { type Story, log_event } from '@storybook/svelte'
import type Component from './Greeting.svelte'

export const Simple: Story<Component> = {
  args: {
    greeting: 'Hello World',
    say_hello: log_event('say_hello')
  },
}
```

With the simplification mentioned at the top of this comment and with Svelte 5's updates, I think standard CSF is looking a lot more attractive for Svelte than it used to be. But that still doesn't solve the use case of component compositions and people who don't want to be bothered by creating a new component just to pass a simple snippet, so Svelte CSF may remain the desire of many. I don't know. Anyhow, keep up the good work and please don't feel the need to reply to any of this. I just wanted to jot down my thoughts.

**Comment by jacob-8** [comment]
> Svelte provides [a set of APIs for mounting a component on the client](https://svelte.dev/docs/client-side-component-api) - this is the API Storybook uses internally to render CSF-based stories. However this API is limited, you can’t do the same things with root Svelte components as you can with components within other components. In practice, this means that you can’t define slots in regular CSF, because there’s no way to pass a [slot](https://svelte.dev/docs/special-elements#slot) to a component when using Svelte’s mounting API, and thus there’s no way for Storybook to take a user’s slot and pass it to the component. The upcoming Svelte 5 introduces [snippets](https://svelte-5-preview.vercel.app/docs/snippets) as a replacement for slots, which have the same limitations.

@JReinhold, snippets do not have the same limitations as slots. Because snippets are just props, they can now be defined and passed in from a regular typescript file (your CSF, or Kitbook's variants/stories file). As long as Storybook does some checking and conversion when necessary, users can use a regular Svelte component as a snippet OR they can use the [`createRawSnippet`](https://svelte-5-preview.vercel.app/docs/imports#svelte-createrawsnippet) api to create a snippet - both are demoed [here](https://svelte-5-preview.vercel.app/#H4sIAAAAAAAACpVVwY6bMBD9lSkbKSChoL2yIdvVSttKVS9tb6VKHBiCVbAt2ySNLP69soGEsNl0qxxij9_zPN6MbeMVtELlxT-Nx0iNXuw9CeGFnj4KO1F7rDR6oad4IzMbWapMUqFXKUs1rQWXGl44h0LyGuaL6IXzRUeaP4wgnySiPoHc7BrMQCaRaPxGDt8ZFQJ1COUxl0QjtD19oEHKYESUyHKUE1SkUO5ROnCqM86Uhq3kB4USErDfyAs4UJbzA3xIEmhYjgVlmKfMEoqGZZpyBlStVafHz3gtOEOm11wOwQCMhadaom4kg2uYRYVsp0tIkgTuLbqd5ECmGonvy0ML8P-lKYAbarptLkVPnPd9IblQ60FgAMkKjDMSAKAz00EggUsoPE4CfgAxmPZhIPcpzTCHvnwx-C7NZpnT_Wpmhlo9wnwOcQ-6-rkhmF5MGyy2PD-2y8jusQnhnEShboSvidw5L0crMLTZm5t3rPCUZERuT-O2G7TBw6jCnVONQrkWku8kqWuiaUaq6rjuTM_XO3c-kreK0Hl_Ua5-ZucXzg3hVC_L-9VnrCoOM-N28YOFPePth2VU3q8G4CYcRs6eGHzGcxxntL8ZFgVm2venC6m28IXGP_qZM43Mfsbmet7Nw5nXe-SG3aBrDxdfRudbhv0oqQKqoOQH56KCA60qYFzW1kMbg75QCigDXSKVQIRYpMzc9SvgHPaNExK4jGd_zIUtJuo5bcqW9m5z1MS4vxYiJ2pZylfivlC95fw3ZISBIEoBgeehm0Bz2KLVmgOxK8O1cJljcgm4izKY5nyqNEpGNN1jdQwv0pJKcSgJy6uRJ32TOW-OAjtrYXvs3WwUZTvr2qvuAyLobYHv6Gon3wu9mue0oJh7sZYNtuHpwXnh_L0PToX2qu8OSwsJzPr-mraMuaNFB3OVNh_712HUBDGk3lfCGqxSzzWEiWjRpuyWUleO_9NqmTelXu_C1yp-tX8BfitUY6wHAAA=).

I'm still considering how to implement in Kitbook but I will probably end up exposing a function that let's the user explicity do the conversion from Component to snippet. Kitbook would provide an `as_snippet` helper:

```ts
import { createRawSnippet, hydrate } from 'svelte'; 
import { render } from 'svelte/server'; 
import { browser } from "$app/environment";
 
export function as_snippet(component: Component) {
  return createRawSnippet((props_function) => { 
    const props = props_function ? props_function() : {};
    return {
      render: () => `<div>${browser ? '' : render(component, { props }).body}</div>`,
      setup(target) { 
        hydrate(component, { target, props }) 
      } 
    }
  }); 
}
```

Then users could do this:

```ts
import { as_snippet, type Story } from 'kitbook'
import type Component from './Greeting.svelte'

import Greet from './Greet.svelte'; // component created just to be used as a snippet

export const SomethingWithASnippet: Story<Component> = {
  props: {
    name: 'Michael',
    greet: as_snippet(Greet)
  },
}
```

Edit: Note that Storybook could ignore the server render part in the `as_snippet` if it only runs client-side.

**Reply by jacob-8** [reply]
https://github.com/sveltejs/svelte/issues/9774 is worth reading for some background though I think the above is sufficient information for our use case of passing snippets to components.

**Comment by MrSuddenJoy** [comment]
Have been building with svelte until they introduced mass surveilance tactics ( = known as telemetry ) that cannot be turned off. This was enough for me!

What this has to do with this project? You are proposing to support CSF made by them. My question is: why to introduce low quality things (svelte) into product of such high quality as `StoryBook`? Honestly I fail to see reason.
